File: CryptoBridge-python/src/infrastructure/external_apis/coinbase_client.py
import asyncio
from typing import List, Optional, Dict, Any
import httpx
from ...domain.entities.ticker import Ticker


class CoinbaseClient:
    BaseUrl = "https://api.exchange.coinbase.com/products"

    def __init__(self):
        self.timeout = httpx.Timeout(10.0, connect=5.0)

    async def getSymbols(self) -> List[str]:
        try:
            async with httpx.AsyncClient(timeout=self.timeout) as client:
                response = await client.get(self.BaseUrl)

                if response.status_code != 200:
                    return []

                data = response.json()
                return [product["id"] for product in data]

        except Exception as e:
            print(f"Erro ao buscar símbolos no Coinbase: {e}")
            return []

    async def getTicker(self, symbol: str) -> Optional[Ticker]:
        try:
            async with httpx.AsyncClient(timeout=self.timeout) as client:
                stats_task = client.get(f"{self.BaseUrl}/{symbol}/stats")
                ticker_task = client.get(f"{self.BaseUrl}/{symbol}/ticker")

                stats_response, ticker_response = await asyncio.gather(
                    stats_task, ticker_task, return_exceptions=True
                )

                if (isinstance(stats_response, Exception) or
                    isinstance(ticker_response, Exception) or
                    stats_response.status_code != 200 or
                    ticker_response.status_code != 200):
                    return None

                stats_data = stats_response.json()
                ticker_data = ticker_response.json()

                return Ticker(
                    symbol=symbol,
                    buy=float(ticker_data.get("bid", 0)),
                    sell=float(ticker_data.get("ask", 0)),
                    vol=float(stats_data.get("volume", 0)),
                    low=float(stats_data.get("low", 0)),
                    high=float(stats_data.get("high", 0)),
                    last=float(ticker_data.get("price", 0))
                )

        except Exception as e:
            print(f"Erro ao buscar ticker {symbol} no Coinbase: {e}")
            return None

File: CryptoBridge-node/src/infrastructure/external-apis/CoinbaseClient.js
const axios = require('axios');
const Ticker = require('../../domain/entities/Ticker');

class CoinbaseClient {
    constructor() {
        this.BaseUrl = 'https://api.exchange.coinbase.com/products';
        this.timeout = 10000; // 10 seconds
    }

    async getSymbols() {
        try {
            const response = await axios.get(this.BaseUrl, {
                timeout: this.timeout
            });

            if (response.status !== 200) {
                return [];
            }

            return response.data.map(product => product.id);
        } catch (error) {
            console.error('Erro ao buscar símbolos no Coinbase:', error.message);
            return [];
        }
    }

    async getTicker(symbol) {
        try {
            const [statsResponse, tickerResponse] = await Promise.all([
                axios.get(`${this.BaseUrl}/${symbol}/stats`, { timeout: this.timeout }),
                axios.get(`${this.BaseUrl}/${symbol}/ticker`, { timeout: this.timeout })
            ]);

            if (statsResponse.status !== 200 || tickerResponse.status !== 200) {
                return null;
            }

            const statsData = statsResponse.data;
            const tickerData = tickerResponse.data;

            return new Ticker(
                symbol,
                parseFloat(tickerData.bid || 0),
                parseFloat(tickerData.ask || 0),
                parseFloat(statsData.volume || 0),
                parseFloat(statsData.low || 0),
                parseFloat(statsData.high || 0),
                parseFloat(tickerData.price || 0)
            );
        } catch (error) {
            console.error(`Erro ao buscar ticker ${symbol} no Coinbase:`, error.message);
            return null;
        }
    }
}

module.exports = CoinbaseClient;